notebooks/examples/python/Ship2ShipTransfers/03. Aggregates.py

```diff
@@ -0,0 +1,204 @@
+spark.conf.set("spark.databricks.labs.mosaic.index.system", "H3")
+mos.enable_mosaic(spark, dbutils)
+
+# COMMAND ----------
+
+cargos_indexed = spark.read.table("ship2ship.cargos_indexed").repartition(
+    sc.defaultParallelism * 20
+)
+display(cargos_indexed)
+cargos_indexed.count()
+
+# COMMAND ----------
+
+# MAGIC %md ## Create Lines
+# MAGIC
+# MAGIC We can `groupBy` across a timewindow to give us aggregated geometries to work with.
+# MAGIC
+# MAGIC When we collect the various points within a timewindow, we want to construct the linestring by the order in which they were generated (timestamp).
+# MAGIC We choose a buffer of a max of 200 metres in this case.
+# MAGIC Since our projection is not in metres, we convert from decimal degrees. With `(0.00001 - 0.000001)` as being equal to one metre at the equator
+# MAGIC Ref: http://wiki.gis.com/wiki/index.php/Decimal_degrees
+
+# COMMAND ----------
+
+lines = (
+    cargos_indexed.repartition(sc.defaultParallelism * 20)
+    .groupBy("mmsi", window("BaseDateTime", "15 minutes"))
+    # We link the points to their respective timestamps in the aggregation
+    .agg(collect_list(struct(col("point_geom"), col("BaseDateTime"))).alias("coords"))
+    # And then sort our array of points by the timestamp to form a trajectory
+    .withColumn(
+        "coords",
+        expr(
+            """
+        array_sort(coords, (left, right) -> 
+            case 
+            when left.BaseDateTime < right.BaseDateTime then -1 
+            when left.BaseDateTime > right.BaseDateTime then 1 
+            else 0 
+            end
+        )"""
+        ),
+    )
+    .withColumn("line", mos.st_makeline(col("coords.point_geom")))
+    .cache()
+)
+
+# COMMAND ----------
+
+# DBTITLE 1,Note here that this decreases the total number of rows across which we are running our comparisons.
+lines.count()
+
+# COMMAND ----------
+
+
+one_metre = 0.00001 - 0.000001
+buffer = 200 * one_metre
+
+
+def get_buffer(line, buffer=buffer):
+    """Create buffer as function of number of points in linestring
+    The buffer size is inversely proportional to the number of points, providing a larger buffer for slower ships.
+
+    Args:
+        line (geometry): linestring geometry as generated with st_makeline.
+
+    Returns:
+        double: buffer size in degrees
+    """
+    np = expr(f"st_numpoints({line})")
+    max_np = lines.select(max(np)).collect()[0][0]
+    return lit(max_np) * lit(buffer) / np
+
+
+cargo_movement = (
+    lines.withColumn("buffer_r", get_buffer("line"))
+    .withColumn("buffer_geom", mos.st_buffer("line", col("buffer_r")))
+    .withColumn("buffer", mos.st_astext("buffer_geom"))
+    .withColumn("ix", mos.mosaic_explode("buffer_geom", lit(9)))
+)
+
+(cargo_movement.createOrReplaceTempView("ship_path"))
+
+display(spark.read.table("ship_path"))
+
+# COMMAND ----------
+
+to_plot = spark.read.table("ship_path").select("buffer").limit(3_000)
```

Comment:
You are plotting the original buffer shape, but it is duplicated for each index. You should be either plotting the indexed chips, or the shapes from the original dataset. 
Once you remove the original shapes from the indexed dataset, you won't have this column available

Reply:
This is purely used as a visual tool to give some intuition about the data—as referenced in the section header below  `# DBTITLE 1,Example Buffer Paths`.

Reply:
Ok, in that case there is only one gotcha: the limit 3000 is applied to the exploded dataset, so each `buffer` geometry is being plotted multiple times. I would add a `distinct()` so that you actually see the 3000 shapes on the map.

Reply:
Good point!